File: Data Structures & Algorithms/count-connected-components/submission-1.cpp
class Solution {
public:
    vector<bool> visited;
    vector<vector<int>> graph;

    void dfs(int u) {
        visited[u] = true;
        for (auto& v: graph[u]) {
            if (!visited[v]) {
                dfs(v);
            }
        }
    }

    int countComponents(int n, vector<vector<int>>& edges) {
        if (n == 1) {
            return 1;
        }
        graph.resize(n);
        //visited = vector<bool>(n);
        visited.resize(n, false);
        for (int i = 0; i < n; i++) {
            graph[i] = vector<int>(n);
        }
        for (auto& edge: edges) {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }
        int count = 0;
        for (int i = 0; i < n; i++) {
            if(!visited[i]) {
                dfs(i);
                count++;
            }
        }
        return count;
    }
};

File: Data Structures & Algorithms/valid-tree/submission-1.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if len(edges) != n - 1:
            return False
        p = [x for x in range(n)]
        sz = [1 for _ in range(n)]
        self.count = n
        # use edge and disjoint set data structure
        def find(x):
            while x != p[x]:
                p[x] = p[p[x]]
                x = p[x]
            return x
        def union(u, v):
            pu = find(u)
            pv = find(v)
            if pu == pv:
                return
            self.count -= 1
            if sz[pu] <= sz[pv]:
                # pu smaller tree, make parent(pu) = pv
                p[pu] = pv
                sz[pv] += sz[pu]
            else:
                p[pv] = pu
                sz[pu] += sz[pv]
        for (u, v) in edges:
            union(u, v)
        return self.count == 1